### PR TITLE
[RFC][http/2 client] Do not reset recently closed streams

### PR DESCRIPTION
**Pull request by deweerdt**
In the current code base, the following can happen:

1) h2o h2 client sends HEADERS
2) h2o h2 client sends RST_STREAM
3) server sees HEADERS, processes it, and sends a HEADERS back
4) h2o h2 client sees the HEADERS belonging to a closed stream and closes
   the entire connection: https://github.com/h2o/h2o/blob/326ab4671639f9321a2288fbb8735411c320420f/lib/common/http2client.c#L974-L980

Attached is a PoC, but it's unsatisfying:
- reset streams are kept forever
- the number of streams to record might be more than the 10 i've hardcoded.

Better options seem to include:
- add a timer and periodically reset `recently_rst_streams`
- sending a PING frame along with the reset, and clear the stream when receiving the PONG

But before embarking in this i figured i would ask for opinions.

### CODE REVIEW

File: t/Util.pm [hunk 1]
```diff
@@ -587,14 +625,22 @@ sub spawn_forked {
...
             kill 'TERM', $pid;
             while (waitpid($pid, 0) != $pid) {}
-        });
+            undef $pid;
+        };
         return +{
             pid => $pid,
-            kill => sub { undef $guard; },
+            kill => $kill,
+            guard => make_guard(sub { $kill->() }),
```
Comment (kazuho) [inline t/Util.pm:643]: Could you please elaborate why we need to make these changes?

IIUC, it is already possible with `spawn_forked` to wait for the child process to die then read stdout, stderr. You can do that by first calling `$child->{kill}->()` (which sends SIGTERM then waits for the process to exit), then reading from `$child->{stdout}` or `$child->{stderr}`.

To me it seems that `on_exit` callback is used for testing output after the child process exits. If that is the case, I think the existing mechanism should be sufficient.

**Review by kazuho — COMMENTED** [review]
Thank you for the fixing this complex problem.

I think the logic is good, I only have nitpicks for the tests.

File: t/Util.pm (hunk 1 above)
Comment (deweerdt) [reply t/Util.pm]: Fixed in e4b279455cec176b4b1af16e7d4415b2036a3e4e, thanks for the advice

File: t/Util.pm [hunk 2]
```diff
@@ -567,6 +569,17 @@ sub wait_debugger {
...
+sub wait_tcp_listen_port {
+    my ($port, $timeout) = @_;
+    while ($timeout-- != 0) {
+        my $out = `ss -tlnp | grep ":$port" 2>&1`;
+        return 1 if (scalar(split(/\n/, $out)) == 1);
+        sleep 1;
+    }
+    print STDERR "Failed to wait for TCP port $port\n";
+    undef;
+}
```
[the remark this answers is not in the document]
Comment (deweerdt) [reply t/Util.pm]: Fixed in 6de9246aeb78753f852031a90bea82f2bdcd99a2

**Comment by kazuho** [comment]
Thank you for the changes! Merged.

**Pull request merged**